### PR TITLE
updated the data structure and added yolo train

### PR DESCRIPTION
**Pull request by mreda1998**
Training of the YOLO architecture imported from ultralytics https://docs.ultralytics.com/datasets/detect/#which-datasets-are-supported-by-ultralytics-yolo-for-object-detection

- Updated the preprocessing script to adapt the path of the processed images for YOLO training
- Labels have also been updated to follow the structure of YOLO object detection

### CODE REVIEW

File: src/models.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import torch
+from ultralytics import YOLO
+import torch.nn as nn
+
+class AxonSegmentationModel(YOLO):
+    def __init__(self, model_path=None):
+        super().__init__(model_path)
+
+        self.input_channels = 3  # Change if your preprocessed images have different channels (e.g., grayscale)
```
Comment (hermancollin) [inline src/models.py:9]: Aren't our images grayscale?

Comment (mreda1998) [reply src/models.py:9]: Yes, finally I did not use this file for the yolo training, it was simply a template if future implementaiton is required. I will update the PR to remove unecessary stuff

File: src/preprocessing.py [hunk 2]
```diff
@@ -15,8 +16,9 @@ def preprocess_data():
     # TODO: Issue of raw path
     data_repo_url = "https://github.com/axondeepseg/data_axondeepseg_sem"
     data_dir = "data_axondeepseg_sem"  # Local directory for the cloned repository
-    processed_images_dir = "processed_images"
-    processed_masks_dir = "processed_masks"
+    processed_images_dir = "dataset/images/train"
+    processed_masks_dir = "dataset/labels/train"
```
Comment (hermancollin) [inline src/preprocessing.py]: Reminder to split the data into train-val-test splits.

File: src/preprocessing.py [hunk 3]
```diff
@@ -53,13 +57,23 @@ def preprocess_data():
...
+                    # cv2.rectangle(img, (minc, minr), (maxc, maxr), (0, 255, 0), 2)  # Green rectangle for myelin
+                    bbox_data.append({"image_name": f"{subject}_{sample}.png", "xmin": minc, "ymin": minr, "xmax": maxr, "ymax": maxc, "class": "myelin"})
+
+                    # Normalize coordinates
+                    img_height, img_width = img.shape[:2]
+                    centroidx /= img_width
+                    centroidy /= img_height
+                    width /= img_width
+                    height /= img_height
+                    file.write('0 {} {} {} {}\n'.format(centroidx, centroidy, width, height))
```
Comment (hermancollin) [inline src/preprocessing.py]: Why do we need a `0` at the start of every line?

File: src/preprocessing.py [hunk 4]
```diff
@@ -53,13 +57,23 @@ def preprocess_data():
...
-                bbox_data.append({"image_name": f"{subject}_{sample}.png", "xmin": minc, "ymin": minr, "xmax": maxr, "ymax": maxc, "class": "myelin"})
+            with open(os.path.join(processed_masks_dir, label_name), "w") as file:
+                for i, region in enumerate(myelin_seg_regions):
+                    minr, minc, maxr, maxc = region.bbox
+                    centroidx, centroidy = region.centroid[0], region.centroid[1]
+                    width,height = region.axis_major_length, region.axis_minor_length
+                    # cv2.rectangle(img, (minc, minr), (maxc, maxr), (0, 255, 0), 2)  # Green rectangle for myelin
+                    bbox_data.append({"image_name": f"{subject}_{sample}.png", "xmin": minc, "ymin": minr, "xmax": maxr, "ymax": maxc, "class": "myelin"})
+
+                    # Normalize coordinates
```
Comment (hermancollin) [inline src/preprocessing.py]: Ok so we are normalizing the bounding box coordinates. I guess this is the format that YOLO expects. Could you indicate clearly the target data format (maybe in the function's docstring)? i.e. is it COCO, LVIS, etc.

File: src/preprocessing.py [hunk 5]
```diff
@@ -53,13 +57,23 @@ def preprocess_data():
...
+                    centroidx /= img_width
+                    centroidy /= img_height
+                    width /= img_width
+                    height /= img_height
+                    file.write('0 {} {} {} {}\n'.format(centroidx, centroidy, width, height))
+
 
-            # Save the image with bounding boxes
-            image_name = f"{subject}_{sample}.png"
             cv2.imwrite(os.path.join(processed_images_dir, image_name), img)
```
Comment (hermancollin) [inline src/preprocessing.py]: Ok so here, you are writing `img` in `dataset/images/train`. These images have bounding boxes overlaid on them. Is this intentional? These images are great for visual inspection of the data, but if we fed this to the detection model we would effectively be "cheating", wouldn't we?

**Review by hermancollin — CHANGES_REQUESTED** [review]

File: src/preprocessing.py (hunk 3 above)
Comment (hermancollin) [reply src/preprocessing.py]: ah nevermind that's the class ID

File: src/preprocessing.py (hunk 2 above)
Comment (mreda1998) [reply src/preprocessing.py]: Corrected in the last commit

File: src/preprocessing.py (hunk 4 above)
Comment (mreda1998) [reply src/preprocessing.py]: resolved in the last commit

File: src/preprocessing.py (hunk 5 above)
Comment (mreda1998) [reply src/preprocessing.py]: Also corrected

**Review by hermancollin — APPROVED** [review]

**Review by hermancollin — APPROVED** [review]
